torch/jit/__init__.py

```diff
@@ -1108,13 +1108,42 @@ def _enable_recursive_script():
     torch._C._jit_recursive_script(False)
 
 
+def script_module_decorator(module_class):
+    """
+    Wraps the `__new__` of a class so that construction always returns a
+    ScriptModule
+    """
+    original_new = getattr(module_class, '__new__', lambda self: None)
+
+    # We wrap `__new__` so that we can intercept instantiations of `module_class`
+    # and wrap them with `torch.jit.script` calls
+    @functools.wraps(original_new)
+    def torchscript__new__wrapper(cls, *args, **kwargs):
+        # Run original new
+        nn_module = original_new(cls)
+
+        # Since we don't return the module itself from this wrapper, its `__init__`
+        # is never automatically called, so we have to do it explicitly here
+        nn_module.__init__(*args, **kwargs)
```

Comment:
I do not think is not correct to make `__new__` call `__init__` and will break assumptions in other parts of python (e.g. things that try to copy the object by calling `__new__`).

Reply:
The point of the decorator is that it compiles your module automatically, which it can't do if it hasn't been initialized. Does it make sense to break those cases since they wouldn't work with `ScriptModule`s anyways?

Reply:
maybe? What do people typically do in class decorators. It seems like there is at least some precedent because it doesn't all __init__ if another object is returned.

Reply:
Class decorators seem pretty rare, the associated [PEP](https://www.python.org/dev/peps/pep-3129/) is also light on details. Most usages seem to be people returning some kind of wrapper class type from the decorator, but returning anything other than the class type itself breaks `super(T, self)` calls (they have to be changed to `super(type(self), self)` which makes that a non-starter for us.

Reply:
Why not wrap the `__init__` instead?